source/common/shared_pool/shared_pool.h

```diff
@@ -61,10 +68,10 @@ class ObjectSharedPool : public Singleton::Instance,
   std::shared_ptr<T> getObject(const T& obj) {
     ASSERT(std::this_thread::get_id() == thread_id_);
     auto hashed_value = HashFunc{}(obj);
-    auto object_it = object_pool_.find(hashed_value);
-    if (object_it != object_pool_.end()) {
-      auto lock_object = object_it->second.lock();
-      if (lock_object) {
+    auto object_range = object_pool_.equal_range(hashed_value);
```

Comment:
rather than emulating a hash-table here by getting a 'bucket' and then iterating through it, can we just use an absl::flat_hash_map in the first place?

Reply:
Could you elaborate on what the `absl::flat_hash_map` is for? What is the key and what is the value for this? 

This is my current understanding of the issue:
The object_pool is currently an `absl::flat_hash_map`, but the keys in that map are non-unique for the case when `HashFunc` produces duplicate values. This can be solved by replacing the `absl::flat_hash_map` with a  multimap that allows for non-unique keys.

Reply:
I see why this is a little tricky, and why you went with a mutlimap.

How about using something similar to the Stats::AllocatorImpl pattern? Use absl::flat_hash_set<T*, HashSharedPtr<T>, CompareSharedPtr<T>>. Clean the set on the shared_ptr's delete callback.

Reply:
The nice thing about `ObjectSharedPool` as it's currently implemented is that it doesn't use locks, as the assumption is that everything is allocated  on the same (main) thread. It does however have a mechanism to schedule deletions of the container values on the original (main) thread. 

I will attempt to keep this behaviour with a `flat_hash_set`.

Reply:
Yeah definitely. I think that should be do-able. If there's something blocking you I'm OK with the approach you have now, but fundamentally I think this can just be a set :)